packages/data-point/src/reducer-helpers/constant/constant.js

```diff
@@ -0,0 +1,30 @@
+const cloneDeep = require("lodash/cloneDeep");
+
+const { Reducer } = require("../../Reducer");
+
+class ReducerConstant extends Reducer {
+  constructor(spec) {
+    super("constant", undefined, spec);
+    // We clone the source to truly respect the nature of a constant, if the
+    // original source is ever changed the data integrity of the constant
+    // should be kept
+    this.constantValue = cloneDeep(spec);
+  }
+
+  static create(spec) {
+    return new ReducerConstant(spec);
+  }
+
+  /**
+   * @returns {any} deep cloned version of the original source
+   */
+  resolve() {
+    // this clone happens so any future changes to the output do not
+    // affect the original value of this.constantValue
+    return cloneDeep(this.constantValue);
```

Comment:
If we need to use cloneDeep for constants, we just shouldn't use constants. A function is easier, and by default it creates a new object for each call.

❌ ```constant({ /* some nested object */ })```

✅ ```const constant = () => /* return some nested object */```

Reply:
we need the cloneDeep here to prevent side effects if the value passed to it suffers any changes afterwards, I do agree a function is faster and prob removes the need of this reducer.

```js
const config = {}

const await result datapoint.resolve(input, [{
  test: constant(config)
}, (value) => {
  // this already not recommended but could happen
  // and would actually change the top level config variable
   value.test.mutate = true
}])
```

I would say we keep this for now, and do a performance check, I think its worth it having it to provide better data integrity.  from the tests I saw on cloneDeep it is pretty solid and soon there will be native support on node v12 or later? https://nodejs.org/api/v8.html#v8_v8_serialize_value

maybe the word `constant` is what's wrong, and changing it to `value` would have less weight?

Reply:
about primitive values cloneDeep should take care of ignoring primitive values, I didn't find it necessary to add that check on our side https://github.com/lodash/lodash/blob/4.17.15/lodash.js#L2632

Reply:
I know that cloneDeep will check the type, but why call it repeatedly for primitives that don't even change? if we keep it, maybe we can have something like this in the constructor:

```js
// just a for example
if (isObject(value)) {
  this.resolve = () => cloneDeep(value)
} else {
  this.resolve = () => value
}
```

Reply:
I like this suggestion

Reply:
> maybe the word constant is what's wrong, and changing it to value would have less weight?

changing the name will not make it execute faster :)

https://jsperf.com/number-clonedeep

Reply:
since I already have a couple of PRs open and isPlainObject is already coded outside of this PR I will create a ticket and follow up on it for refactoring once these PRs get merged

Reply:
https://github.com/ViacomInc/data-point/issues/386

about changing the name I meant to not be a `constant` so we don't have to adhere to the meaning of the _word_ "constant", so no need to add any hooks to make sure value passed is "constant" and instead it will behave as a reference (in the case of an object being passed)

Reply:
ohhh, I see. that makes sense.